cmd/login.go

```diff
@@ -11,40 +13,73 @@ import (
 	"golang.org/x/crypto/ssh/terminal"
 )
 
-func init() {
-	var (
-		loginCmd = &cobra.Command{
-			Use:   "login",
-			Short: "Login to AppLand",
-			Args:  cobra.MaximumNArgs(1),
-			Run: func(cmd *cobra.Command, args []string) {
-				reader := bufio.NewReader(os.Stdin)
-				context, err := config.GetCurrentContext()
-				if err != nil {
-					fail(err)
-				}
+type PasswordReader func() ([]byte, error)
+
+var stdinPasswordReader = func() ([]byte, error) {
+	return terminal.ReadPassword(0)
+}
+
+func NewLoginCommand(connecter Connecter, stdin io.Reader, passwordReader PasswordReader) *cobra.Command {
+	return &cobra.Command{
+		Use:   "login",
+		Short: "Login to AppLand",
+		Args:  cobra.MaximumNArgs(1),
+		PersistentPreRun: func(cmd *cobra.Command, args []string) {
+			config.LoadCLIConfig()
+			// defer connecting to the server until Run (below)
+		},
+		Run: func(cmd *cobra.Command, args []string) {
+			reader := bufio.NewReader(stdin)
 
-				fmt.Printf("logging into %s\n\n", context.GetURL())
-				fmt.Printf("login: ")
-				login, err := reader.ReadString('\n')
+			api := connecter()
+
+			context, err := config.GetCurrentContext()
+			if err != nil {
+				fail(err)
+			}
+
+			fmt.Printf("logging into %s\n\n", context.GetURL())
+			fmt.Printf("login: ")
+			login, err := reader.ReadString('\n')
+			if err != nil {
+				fail(err)
+			}
+			login = strings.TrimSpace(login)
+
+			validAPIKey := false
+			data, err := base64.RawURLEncoding.DecodeString(login)
+			if err == nil && len(strings.Split(string(data), ".")) == 2 {
+				// Looks like an API key, try it out
+				validAPIKey, err = api.TestAPIKey(login)
 				if err != nil {
 					fail(err)
 				}
+				if !validAPIKey {
+					fail(fmt.Errorf("Invalid API key"))
+				}
+			}
 
+			if validAPIKey {
+				context.SetAPIKey(login)
+			} else {
 				fmt.Printf("password: ")
-				password, err := terminal.ReadPassword(0)
+				bytes, err := passwordReader()
 				if err != nil {
 					fail(err)
 				}
-				fmt.Printf("\n\nlogged in.\n")
 
-				err = api.Login(strings.TrimSpace(login), strings.TrimSpace(string(password)))
+				fmt.Println("logging in....")
```

Comment:
Four dots look weird.... :)

Reply:
I agree. I didn't actually change this line -- the changes I made around it tricked GH into thinking I did.

Reply:
Fair. OTOH I suppose then there's no better time to fix it, seeing it won't dirty the diff now!